cpp/log/cert.cc

```diff
@@ -43,6 +43,99 @@ static int X509_get_signature_nid(const X509* x) {
 #endif
 
 
+namespace {
+#if defined(OPENSSL_IS_BORINGSSL)
+// BoringSSL doesn't have DSA hooked up so to accept these certs we have
+// to do the work ourselves. Can be called with a non DSA issuer key but will
+// always return false in that case.
+//
+// cert is the certificate that is to be checked
+// issuer_key is the public key of the certificate issuer (should be DSA)
+// If dsa_key_with_params is not null it's a key that can be used to copy
+// the DSA parameters if they are absent from the issuer key.
+// Returns true if the DSA signature was correctly verified or false if
+// it did not or there was any error along the way.
+StatusOr<bool> check_dsa_signature(const X509* cert, EVP_PKEY* issuer_key,
+                                   const EVP_PKEY* dsa_key_with_params) {
+  // Before we start rummaging about through pointers ensure everything we need
+  // is present
+  if (!cert || !issuer_key || !cert->cert_info || !cert->cert_info->signature ||
```

Comment:
Avoid reaching to struct internals where possible. I don't know what, if anything, BoringSSL will make opaque here but  nevertheless. Bssl has accessors for `X509_get_cert_info` etc.